src/modules/hip/kernel/tensor_sum.hpp

```diff
@@ -0,0 +1,1433 @@
+{
+    int id_x = hipThreadIdx_x * 8;
+    int id_z = hipBlockIdx_z;
+
+    __shared__ uint partialSum_smem[1024];                          // 8192 uints of src reduced to 1024 in a 1024 x 1 thread block
+    partialSum_smem[hipThreadIdx_x] = 0;                            // initialization of Shared to 0 using all 1024 x 1 threads
+
+    if (id_x >= xBufferLength)
+    {
+        return;
+    }
+
+    int xAlignedLength = xBufferLength & ~7;                        // alignedLength for vectorized global loads
+    int xDiff = xBufferLength - xAlignedLength;                     // difference between bufferLength and alignedLength
+    uint srcIdx = (id_z * xBufferLength) + id_x;
+
+    d_uint8 src_ui8;
+    *(d_uint8_s *)(&src_ui8) = *(d_uint8_s *)(srcPtr + srcIdx);
+
+    if (id_x + 8 > xBufferLength)
+        for(int i = xDiff; i < 8; i++)
+            src_ui8.ui1[i] = 0;                                     // local memory reset of invalid values (from the vectorized global load) to 0
+
+    src_ui8.ui4[0] += src_ui8.ui4[1];                               // perform small work of vectorized uint4 addition
+    partialSum_smem[hipThreadIdx_x] += (src_ui8.ui1[0] +
+                                        src_ui8.ui1[1] +
+                                        src_ui8.ui1[2] +
+                                        src_ui8.ui1[3]);            // perform small work of reducing uint32s to uint using 1024 x 1 threads and store in Shared
+    __syncthreads();                                                // syncthreads after Shared load
+
+    // Reduction of 1024 uints on 1024 threads per block in x dimension
+    for (int threadMax = 512; threadMax >= 1; threadMax /= 2)
+    {
+        if (hipThreadIdx_x < threadMax)
+            partialSum_smem[hipThreadIdx_x] += partialSum_smem[hipThreadIdx_x + threadMax];
+        __syncthreads();
+    }
+
+    // Final store to dst
+    if (hipThreadIdx_x == 0)
+        dstPtr[hipBlockIdx_z] = (Rpp64u)partialSum_smem[0];
+}
+
+__global__ void tensor_sum_grid_result(Rpp32s *srcPtr,
+                                       uint xBufferLength,
+                                       Rpp64s *dstPtr)
+{
+    int id_x = hipThreadIdx_x * 8;
+    int id_z = hipBlockIdx_z;
+
+    __shared__ int partialSum_smem[1024];                           // 8192 floats of src reduced to 1024 in a 1024 x 1 thread block
+    partialSum_smem[hipThreadIdx_x] = 0;                            // initialization of Shared to 0 using all 1024 x 1 threads
+
+    if (id_x >= xBufferLength)
+    {
+        return;
+    }
+
+    int xAlignedLength = xBufferLength & ~7;                        // alignedLength for vectorized global loads
+    int xDiff = xBufferLength - xAlignedLength;                     // difference between bufferLength and alignedLength
+    uint srcIdx = (id_z * xBufferLength) + id_x;
+
+    d_int8 src_i8;
+    *(d_int8_s *)(&src_i8) = *(d_int8_s *)(srcPtr + srcIdx);
+    if (id_x + 8 > xBufferLength)
```

Comment:
this code is ambiguous. What is this trying to do?

Reply:
This will help reset invalid values to 0.
Towards right end of the image when width isn't divisible by bufferlength i.e 8 when we could have few junk values, so this segment of code is used to reset those values.

Reply:
This will help reset invalid values to 0.
Towards right end of the image when width isn't divisible by bufferlength i.e 8, we could have few junk values, so this segment of code is used to reset those values.

Reply:
I am not sure if this is the right way to do this having a loop. we need to make the load to load valid pixels instead.
todo:: revisit this code later